### DISCUSSION TITLE
How to dynamically change DataGridTextColumn from code using mvvm?

### DISCUSSION THREAD

**Discussion by vo1s** [AvaloniaUI/Avalonia#12953] [Q&A] [answered]
I have 4 different tables with different amount of columns and their names. So depending on combobox selection i need to display the necessary one. I need it to be one DataGrid with dynamically changing columns instead of 4 different DataGrid's.
I guess it has smth to do with
 `public ObservableCollection<DataGridColumn> Columns { get; }`
But I don't understand how to set it in code.

**Comment by timunie** [comment]
`MyDataGrid.Columns.Add()  ... .Remove()`

or `MyDataGrid.Columns[0].IsVisible=True|False;`

**Reply by vo1s** [reply]
I cannot access MyDataGrid.Columns. Because I use Binding. I need to work with ObservableCollection through code

**Reply by timunie** [reply]
can't you bind IsVsible to bool per column?

**Reply by vo1s** [reply]
so I put every column in DataGrid and just manage their visibility?

**Reply by timunie** [reply]
yeah or you need to add / remove them as I told you earlier. Its not bindable. You need Interaction or a MessageBus to do. So I prefer binding to IsVisible if I expect only a hand full columns

**Comment by workgroupengineering** [comment]
Columns  is not Bindable. You can create custom Behaviors like this:

```csharp
internal class DataGridBindableColumns : Behavior<DataGrid>
{

    public static readonly AvaloniaProperty<ObservableCollection<DataGridColumn>?> ColumnsProperty =
        AvaloniaProperty.Register<DataGridBindableColumns, ObservableCollection<DataGridColumn>?>(nameof(Columns));

    private readonly System.Collections.Specialized.NotifyCollectionChangedEventHandler SyncColumns;

    public DataGridBindableColumns()
    {
        SyncColumns = SyncColumnsImp;
    }

    public ObservableCollection<DataGridColumn>? Columns
    {
        get => this.GetValue<ObservableCollection<DataGridColumn>?>(ColumnsProperty);
        set => SetValue(ColumnsProperty, value);
    }

    protected override void OnAttachedToVisualTree()
    {
        base.OnAttachedToVisualTree();
        RefreshColumns(Columns);
    }

    private void RefreshColumns(ObservableCollection<DataGridColumn>? columns)
    {
        AssociatedObject!.Columns.Clear();
        if (columns is not null)
        {
            foreach (DataGridColumn column in columns)
            {
                AssociatedObject!.Columns.Add(column);
            }
            columns.CollectionChanged += SyncColumns;
        }
    }

    private void SyncColumnsImp(object? sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
    {
        switch (e.Action)
        {
            case System.Collections.Specialized.NotifyCollectionChangedAction.Add:
                //TODO : Add
                break;
            case System.Collections.Specialized.NotifyCollectionChangedAction.Remove:
                //TODO : Remove
                break;
            case System.Collections.Specialized.NotifyCollectionChangedAction.Replace:
                //TODO : Replace
                break;
            case System.Collections.Specialized.NotifyCollectionChangedAction.Move:
                //TODO : Move
                break;
            case System.Collections.Specialized.NotifyCollectionChangedAction.Reset:
                //TODO : Reset
                break;
            default:
                break;
        }
    }

    protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change)
    {
        base.OnPropertyChanged(change);
        if (change.Property == ColumnsProperty)
        {
            if (change.OldValue is INotifyCollectionChanged oldNotify)
            {
                oldNotify.CollectionChanged -= SyncColumns;
            }
            var columns = change.NewValue as ObservableCollection<DataGridColumn>;
            RefreshColumns(columns);
        }
    }
}
```

and use

```xaml

  <DataGrid>
    <Interaction.Behaviors>
      <l:DataGridBindableColumns Columns="{Binding Columns}"/>
    </Interaction.Behaviors>
  </DataGrid>

```

don't forget to add the package `Avalonia.Xaml.Behaviors`

**Reply by vo1s** [reply]
Thank you!

**Reply by vo1s** [reply]
I've managed to display columns from observableCollection, but now the information does not appear because a I need to bind them like this:
<DataGridTextColumn Width="58" Header="№" Binding=**"{Binding KrnListValueId}**"></DataGridTextColumn>
Thats how I create columns:
```
[ObservableProperty]
    private ObservableCollection<DataGridColumn>? _columns = new ObservableCollection<DataGridColumn>()
    {
        new DataGridTextColumn(){Header = "1", Binding = new Binding()},
        new DataGridTextColumn()Header = "2", Binding = new Binding()},
        new DataGridTextColumn(){Header = "3", Binding = new Binding()}
    };
```
So how do i bind it to the specific column like **KrnListValueId**

KrnListValueId is a column from data base. I am using EF and have it in my Models:
```
[Table("krn_list_value")]
public partial class KrnListValue
{
    [Key]
    [Column("krn_list_value_id", TypeName = "int(11)")]
    public int KrnListValueId { get; set; }

    [Column("krn_list_id", TypeName = "int(11)")]
    public int? KrnListId { get; set; }
```

**Reply by workgroupengineering** [reply, marked as answer]
```c#
[ObservableProperty]
    private ObservableCollection<DataGridColumn>? _columns = new ObservableCollection<DataGridColumn>()
    {
        new DataGridTextColumn(){Header = "1", Binding = new Binding("KrnListValueId"){ Source = this.KrnListValueList }},
        ....
    };
```

**Reply by vo1s** [reply]
Thank you so much! Works perfect now

**Reply by QuangHaii** [reply]
> ```cs
> [ObservableProperty]
>     private ObservableCollection<DataGridColumn>? _columns = new ObservableCollection<DataGridColumn>()
>     {
>         new DataGridTextColumn(){Header = "1", Binding = new Binding("KrnListValueId"){ Source = this.KrnListValueList }},
>         ....
>     };
> ```

Hello thanks for the nice code you got here but is there anyway to add columns using a button ?
I tried to do a button command like this but it didn't work
```
 [RelayCommand]
 public void Button( )
 {
     Columns.Add(new DataGridTextColumn() { Header = "8", Binding = new Binding() });
 }
```
It only worked if i do this 
```
[RelayCommand]
 public void Button( )
 {
     Columns = new ObservableCollection<DataGridColumn>()
    {
    new DataGridTextColumn(){Header = "1", Binding = new Binding()},
    new DataGridTextColumn(){Header = "2", Binding = new Binding()},
    new DataGridTextColumn(){Header = "3", Binding = new Binding()},
    new DataGridTextColumn(){Header = "4", Binding = new Binding()},
    new DataGridTextColumn(){Header = "5", Binding = new Binding()},
    new DataGridTextColumn() { Header = "8", Binding = new Binding() })
   };
}
```
Is there anyway to add new column to the collections instead of creating a new collection ? Thank you

**Reply by thevortexcloud** [reply]
This approach violates MVVM by directly coupling the view model to the view. You should instead try to handle this in code behind of the view. You can still have the code behind listen to changes on the view model and update the column list as needed, as long as the view model does not have any references to Avalonia code.

**Reply by QuangHaii** [reply]
> This approach violates MVVM by directly coupling the view model to the view. You should instead try to handle this in code behind of the view. You can still have the code behind listen to changes on the view model and update the column list as needed, as long as the view model does not have any references to Avalonia code.

Because my number of columns depend on my number of data so if I want to handle it in code behind then I have to pass data to view and it still violates MVVM in my opinion. Btw, I don't know how to get view model data in code behind.

**Reply by thevortexcloud** [reply]
>  have to pass data to view and it still violates MVVM in my opinion

If the view model is directly passing data to the view, you are correct. However there is no violation if the view reads data from the view model on its own/reacts to a change event from the view model. As such there is nothing stopping you from manually subscribing to a change event on an `ObservableCollection` (probably containing some sort of view model) in your code behind.

> . Btw, I don't know how to get view model data in code behind.

Typically you use the `DataContext` property, which will be set to the instance of your view model. You will need to cast it to your view model type. If you are using a ReactiveUI view you will also have a strongly typed `ViewModel` property.

**Reply by workgroupengineering** [reply]
> > ```cs
> > [ObservableProperty]
> >     private ObservableCollection<DataGridColumn>? _columns = new ObservableCollection<DataGridColumn>()
> >     {
> >         new DataGridTextColumn(){Header = "1", Binding = new Binding("KrnListValueId"){ Source = this.KrnListValueList }},
> >         ....
> >     };
> > ```
> 
> Hello thanks for the nice code you got here but is there anyway to add columns using a button ?
> I tried to do a button command like this but it didn't work
> ```
>  [RelayCommand]
>  public void Button( )
>  {
>      Columns.Add(new DataGridTextColumn() { Header = "8", Binding = new Binding() });
>  }
> ```
> It only worked if i do this 
> ```
> [RelayCommand]
>  public void Button( )
>  {
>      Columns = new ObservableCollection<DataGridColumn>()
>     {
>     new DataGridTextColumn(){Header = "1", Binding = new Binding()},
>     new DataGridTextColumn(){Header = "2", Binding = new Binding()},
>     new DataGridTextColumn(){Header = "3", Binding = new Binding()},
>     new DataGridTextColumn(){Header = "4", Binding = new Binding()},
>     new DataGridTextColumn(){Header = "5", Binding = new Binding()},
>     new DataGridTextColumn() { Header = "8", Binding = new Binding() })
>    };
> }
> ```
> Is there anyway to add new column to the collections instead of creating a new collection ? Thank you

Because,  in Sample, missing `TODO` parts.

**Reply by QuangHaii** [reply]
> Typically you use the `DataContext` property, which will be set to the instance of your view model. You will need to cast it to your view model type

Thank you for your explanation. Basically, for example if I have a `ObservableCollection listData` in my view model then I read it from code behind using `DataContext.listData` and it won't violate MVVM right?

> As such there is nothing stopping you from manually subscribing to a change event on an `ObservableCollection`

How do I subscribe to a change event in my code behind? Do I have to create a new function in code behind then bind it to `DataContext.listData.CollectionChanged` ?

**Reply by QuangHaii** [reply]
> Because, in Sample, missing `TODO` parts.

Sorry because I'm new in Avalonia and I have no idea what to do in the `TODO` parts to fix my current problem.

**Reply by thevortexcloud** [reply]
> Thank you for your explanation. Basically, for example if I have a ObservableCollection listData in my view model then I read it from code behind using DataContext.listData and it won't violate MVVM right?

Correct. The view is allowed to know about the view model. The view model is not allowed to know about the view.

> How do I subscribe to a change event in my code behind? Do I have to create a new function in code behind then bind it to DataContext.listData.CollectionChanged ?

It's just a standard C# event, so use the standard C# event subscription system. If the datacontext changes (which is almost never does in the vast majority of code), you probably also need to handle unsubscribing and resubscribing. You can subscribe in the constructor if your `DataContext` is not null at that point, or you can do it in the `Loaded` event if it gets set slightly later.

You can't really bind to events. Avalonia itself has functionality to listen to `INotifyCollectionChanged` events when you try to bind to a collection that implements it. But there is not much point doing a binding in this situation.

**Reply by QuangHaii** [reply]
> Avalonia itself has functionality to listen to `INotifyCollectionChanged` events when you try to bind to a collection that implements it

I don't understand how this worked. For example, I have a function in code behind to update my column depends on the collection. If my `ObservableCollection` changed in my view model, then how can my view notice it and call the function to update my column?

**Reply by thevortexcloud** [reply]
That's why you subscribe to the change event on the collection. The event is the notification to update the view. It would very roughly (I have not tested this) look like this:

```csharp
((MyViewModel)DataContext).MyCollection.CollectionChanged += (sender,args) => {
MyDataGrid.Columns.Add(new DataGridTextColumn());
}

```

**Reply by QuangHaii** [reply]
Thank you for your explanation and your code sample, I think I understand now.  Thank you for your help once again
